src/main/java/io/cryostat/discovery/PluginInfoDao.java

```diff
@@ -67,78 +67,88 @@ class PluginInfoDao extends AbstractDao<UUID, PluginInfo> {
         this.gson = gson;
     }
 
-    public PluginInfo save(String realm, URI callback, EnvironmentNode subtree) {
-        Objects.requireNonNull(realm);
-        Objects.requireNonNull(subtree);
-        return super.save(new PluginInfo(realm, callback, gson.toJson(subtree)));
+    public final PluginInfo save(String realm, URI callback, EnvironmentNode subtree) {
+        synchronized (entityManager) {
+            Objects.requireNonNull(realm);
+            Objects.requireNonNull(subtree);
+            return super.save(new PluginInfo(realm, callback, gson.toJson(subtree)));
+        }
     }
 
-    public List<PluginInfo> getByRealm(String realm) {
-        Objects.requireNonNull(realm);
+    public final List<PluginInfo> getByRealm(String realm) {
+        synchronized (entityManager) {
```

Comment:
In theory probably not, because the backing storage should be using atomic transactions, so what we get for a get request should either be the complete state before or complete state after any other mutation transaction (like a plugin publishing a new subtree). In this case I don't think there is any risk of deadlocking or other similar structural concurrency issues so it may just reduce throughput somewhat. We can try removing the locking here and see if performance is improved while also still ensuring data consistency and correctness, but I think we should hold onto that experiment until after the 2.2 release branching is done.